### ISSUE TITLE
08-k-nearest-neighbours-winequality-red.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/crerov-intro-machine-learning#6]
@crerov 

### Error 1: Guardas el modelo equivocado

Este es el error más grave. Mira tu código al final:

```python
pickle.dump(model, f)  # ← Estás guardando "model"
```

Pero `model` es el KNN inicial con k=5. Estás guardando el modelo peor.

**Corrección:**
```python
with open("../models/KNN.pkl", 'wb') as f:
    pickle.dump(grid_model, f)  # ← Guardar el modelo optimizado
```

---

### Error 2: No guardas el scaler

Sin el scaler guardado, no puedes usar el modelo en producción. Cuando llegue un vino nuevo, necesitas escalarlo exactamente igual que los datos de entrenamiento.

**Corrección:**
```python
with open("../models/scaler.pkl", 'wb') as f:
    pickle.dump(scaler, f)
```

---

### Error 3: Falta la función predict_wine_quality()

El enunciado pide crear una función que reciba los valores químicos de un vino y devuelva su calidad predicha. Se que dice si te sientes confiado, pero pruebate que le da sentido al trabajo. 

**Corrección:**
```python
def predict_wine_quality(features):
    """
    Recibe una lista con los 11 valores químicos del vino.
    Devuelve un mensaje con la calidad predicha.
    """
    columnas = X.columns.tolist()
    features_df = pd.DataFrame([features], columns=columnas)
    features_scaled = scaler.transform(features_df)
    prediction = grid_model.predict(features_scaled)[0]
    
    if prediction == 2:
        return "Este vino probablemente sea de alta calidad"
    elif prediction == 1:
        return "Este vino probablemente sea de calidad media"
    else:
        return "Este vino probablemente sea de baja calidad"

# Prueba con el ejemplo del enunciado
print(predict_wine_quality([7.4, 0.7, 0.0, 1.9, 0.076, 11.0, 34.0, 0.9978, 3.51, 0.56, 9.4]))
```

---

### Error 4: Falta la gráfica accuracy vs k

El enunciado pide hacer un loop probando valores de k del 1 al 20 y graficar los resultados. Usaste GridSearchCV que es válido, pero no cumple con el requisito de la gráfica visual.

**Corrección (añadir antes del GridSearchCV):**
```python
accuracies = []
k_range = range(1, 21)

for k in k_range:
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train_scaled, y_train)
    accuracies.append(knn.score(X_test_scaled, y_test))

plt.figure(figsize=(10, 6))
plt.plot(k_range, accuracies, marker='o')
plt.title('Accuracy vs. Valor de K')
plt.xlabel('K')
plt.ylabel('Accuracy')
plt.grid(True)
plt.show()

print(f"Mejor k: {k_range[np.argmax(accuracies)]}")
```

---

### Error 5: Faltan métricas de evaluación

Solo muestras accuracy. El enunciado pide también confusion_matrix y classification_report.

**Corrección:**
```python
y_pred_final = grid_model.predict(X_test_scaled)

print("Matriz de Confusión:")
print(confusion_matrix(y_test, y_pred_final))

print("\nReporte de Clasificación:")
print(classification_report(y_test, y_pred_final, 
      target_names=['Baja (0)', 'Media (1)', 'Alta (2)']))
```

---

### Error menor: Falta stratify en train_test_split

Con clases desbalanceadas (muy pocos vinos de calidad baja), es importante mantener las proporciones en train y test.

**Corrección:**
```python
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
```

---

### Error menor: scaler.fit() redundante

Tienes esto:
```python
scaler.fit(X_train)  # ← Innecesario
X_train_scaled = scaler.fit_transform(X_train)  # ← Esto ya hace fit + transform
```

`fit_transform()` ya hace el fit internamente, así que la primera línea sobra.

---

Buen trabajo con el EDA y GridSearchCV, pero faltan requisitos importantes del enunciado. Corrige los puntos anteriores y tendrás un proyecto completo.
